test/proxy/reliabletopic/reliable_topic_test.go

```diff
@@ -0,0 +1,359 @@
+}
+
+func TestReliableTopicProxy_MessageFieldSetCorrectly(t *testing.T) {
+	reliableTopic, _ := client.GetReliableTopic("myReliableTopic")
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	listener := &ReliableMessageListenerMock{wg: wg, storedSeq: -1}
+	id, _ := reliableTopic.AddMessageListener(listener)
+	defer reliableTopic.RemoveMessageListener(id)
+	msg := "foobar"
+	beforePublishTime := time.Now()
+	err := reliableTopic.Publish(msg)
+	afterPublishTime := time.Now()
+	assert.ErrorNil(t, err)
+	timeout := test.WaitTimeout(wg, test.Timeout)
+	assert.Equal(t, err, timeout, false)
+	assert.Equal(t, err, len(listener.messages), 1)
+	assert.Equal(t, err, listener.messages[0].MessageObject(), msg)
+	assert.Equal(t, err, listener.messages[0].PublishingMember(), nil)
+
+	actualPublishTime := listener.messages[0].PublishTime()
+	log.Println(actualPublishTime, " ", beforePublishTime)
+	if actualPublishTime.Second() < beforePublishTime.Second() {
+		t.Error("actualPublish time should be after publish")
+	}
+
+	if afterPublishTime.Second() > afterPublishTime.Second() {
+		t.Error("actualPublish time should be less than afterPublishTime")
+	}
+
+}
+
+func TestReliableTopicProxy_AlwaysStartAfterTail(t *testing.T) {
+	reliableTopic, _ := client.GetReliableTopic("myReliableTopic")
+	reliableTopic.Publish("1")
+	reliableTopic.Publish("2")
+	reliableTopic.Publish("3")
+
+	wg := new(sync.WaitGroup)
+	wg.Add(3)
+	listener := &ReliableMessageListenerMock{wg: wg, storedSeq: -1}
+	id, _ := reliableTopic.AddMessageListener(listener)
+	defer reliableTopic.RemoveMessageListener(id)
+
+	reliableTopic.Publish("4")
+	reliableTopic.Publish("5")
+	reliableTopic.Publish("6")
+
+	timeout := test.WaitTimeout(wg, test.Timeout)
+	assert.Equal(t, nil, timeout, false)
+
+	assert.Equal(t, nil, len(listener.messages), 3)
+	assert.Equal(t, nil, listener.messages[0].MessageObject(), "4")
+	assert.Equal(t, nil, listener.messages[1].MessageObject(), "5")
+	assert.Equal(t, nil, listener.messages[2].MessageObject(), "6")
+}
+
+func TestReliableTopicProxy_Discard(t *testing.T) {
+	reliableTopic, _ := client.GetReliableTopic("discard")
+	topic := reliableTopic.(*internal.ReliableTopicProxy)
+	items := generateItems(client.(*internal.HazelcastClient), 10)
+	topic.Ringbuffer().AddAll(items, core.OverflowPolicyFail)
+	topic.Publish(11)
+	seq, err := topic.Ringbuffer().TailSequence()
+	assert.ErrorNil(t, err)
+	item, err := topic.Ringbuffer().ReadOne(seq)
+	assert.ErrorNil(t, err)
+	msg := item.(*reliabletopic.Message)
+	obj, _ := client.(*internal.HazelcastClient).SerializationService.ToObject(msg.Payload().(*serialization.Data))
+	assert.Equal(t, nil, obj, int64(10))
+}
+
+func TestReliableTopicProxy_Overwrite(t *testing.T) {
+	reliableTopic, _ := client.GetReliableTopic("overwrite")
+	topic := reliableTopic.(*internal.ReliableTopicProxy)
+	items := generateItems(client.(*internal.HazelcastClient), 10)
+	topic.Ringbuffer().AddAll(items, core.OverflowPolicyFail)
```

Comment:
Why don't we use topic.publish for first 10 items ?

Reply:
For this test I think it can be used but what would be the advantage?

Reply:
We are testing the ReliableTopic here. We should use RingBuffer only when necessary. Otherwise it looks confusing to the reader.